### PR TITLE
Add Pure Render Decorators

### PR DESCRIPTION
**Pull request by taion**
For gaearon/react-pure-render#2

Should this be `@PureDecorator` instead?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,8 +2,9 @@ node_modules
...
-index.js
-mixin.js
-shallowEqual.js
\ No newline at end of file
+/component.js
+/decorator.js
+/function.js
+/index.js
+/mixin.js
+/shallowEqual.js
```
Comment (gaearon) [inline .gitignore:10]: Why `/`?

File: src/decorator.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import shouldPureComponentUpdate from './function';
+import { Component } from 'react';
+
+export default function PureRender(Component) {
+  const componentName = Component.name || 'component';
```
Comment (gaearon) [inline src/decorator.js]: Please make it `Component.displayName || Component.name || 'component'`

File: src/decorator.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+  if (!(Component.prototype instanceof Component)) {
+    throw new Error(`${componentName} is not a React component`);
```
Comment (gaearon) [inline src/decorator.js]: This is not necessary. You don't have to inherit from `Component` for it to work. Maybe replace this with a simple check for prototype's existance?

File: .gitignore (hunk 1 above)
Comment (taion) [inline .gitignore:10]: Otherwise it picks up the files under src/. I could `add -f`, but this seems better.

File: src/decorator.js (hunk 3 above)
Comment (taion) [inline src/decorator.js]: Why not? Don't all React components have to inherit from `Component`? It wouldn't make sense to apply this to an ES5-style React component.

File: src/decorator.js (hunk 2 above)
Comment (taion) [inline src/decorator.js]: Will do.

File: .gitignore (hunk 1 above)
Comment (gaearon) [inline .gitignore:10]: Maybe `./` works? `/` seems kinda weird

File: src/decorator.js (hunk 3 above)
Comment (gaearon) [inline src/decorator.js]: > Don't all React components have to inherit from Component?

Nope. You can totally not inherit, as long as you don't need `forceUpdate` or `setState`. I do it all the time.

Also, why not allow it on `createClass`-style components? It should work, no?

File: .gitignore (hunk 1 above)
Comment (taion) [inline .gitignore:10]: This pattern is explicitly documented in the Git book: http://git-scm.com/docs/gitignore

> A leading slash matches the beginning of the pathname. For example, "/*.c" matches "cat-file.c" but not "mozilla-sha1/sha1.c".

And `./` doesn't work - I just checked.

File: src/decorator.js (hunk 3 above)
Comment (taion) [inline src/decorator.js]: Huh. I didn't realize you could do that. Cool! Okay, will make this change too.

**Comment by taion** [comment]
Do you want me to squash these fixes, or would you prefer a separate commit?

File: .gitignore (hunk 1 above)
Comment (gaearon) [inline .gitignore:10]: Oh I see. Ok!

**Comment by gaearon** [comment]
Let's squash

**Comment by gaearon** [comment]
Now that I think of it, there's another approach I like more. Decorating render function. The decorator would return memoized result of props, state and context are the same. 

Can you look into making that work?

**Comment by gaearon** [comment]
See https://mobile.twitter.com/sebmarkbage/status/595336407259811841.

**Comment by taion** [comment]
Huh. Cool! I can take a look at that over the weekend (on a separate PR). Need to look into whether I need to make (shallow) copies of the old props/state objects to make that work.

What are your thoughts regarding naming? Should I rename this to e.g. `@pureClass`? And call the latter `@pureMethod`?

**Comment by taion** [comment]
I incorporated the requested changes.

The new prototype check won't prevent someone from using this to decorate a function or something. Then again, someone decorating something other than a class with `@pureClass` probably deserves whatever they were getting, anyway.

**Comment by taion** [comment]
I added `@pureMethod`. Two questions:

1) I forward `arguments`. This isn't necessary for `#render()`, but might be nice for other methods. If I'm not mistaken, this will inflict a performance hit on most VMs. Do you think this is worth keeping?
2) Cursory testing indicates that I don't need to make a copy of the old `state` and `props`. Do you think I should make a shallow copy just to be safe?

**Comment by gaearon** [comment]
I'm very busy in the beginning of this week so I'm going to take a look a bit later, sorry!

**Comment by taion** [comment]
No worries. In addition to the 2 questions above, it just occurred to me that I could merge `@pureClass` and `@pureMethod` by just checking the signature with which the decorator was called. Let me know if you think that'd be preferable (I don't think it would be, though, since it's much less explicit and for no good reason).

**Comment by taion** [comment]
Did you get a chance to look at this?

**Comment by gaearon** [comment]
@taion Not yet. I might be able in a week or two. Currently busy with a talk I'm giving tomorrow. :-)

**Comment by taion** [comment]
Oh, right. (: Good luck!

**Comment by axyz** [comment]
any news about that?

**Comment by michaelcontento** [comment]
+1

**Comment by taion** [comment]
Ping!

The `@pureMethod` decorator is hackier than I'd like, but I think people would find `@pureClass` to be helpful.

**Comment by gaearon** [comment]
No news yet. I just released Redux 1.0 a few weeks ago and has been busy with React Hot Loader 2 recently which needs a lot of work and has much more priority. When I'm idle, I'll look at this, but for now, feel free to release this separately if I'm being frustrating. Sorry!

**Comment by janusch** [comment]
Is this going to be merged? Would suit your, @gaearon , functional style much better than extending from PureComponent.

**Comment by gaearon** [comment]
Ugh. I didn’t handle this one well. I’m sorry! :disappointed_relieved: 

I decided to deprecate this project, as there are better maintained options now:
https://github.com/gaearon/react-pure-render#no-maintenance-intended

Thank you for sending the PR, and sorry for the trouble!

**Comment by taion** [comment]
No worries.

**Comment by janusch** [comment]
Which alternatives are you thinking of? This one: https://github.com/acdlite/recompose ?

**Pull request closed without merging**